src/modules/keyboardq/KeyboardQmlViewStep.cpp

```diff
@@ -0,0 +1,146 @@
+
+KeyboardQmlViewStep::KeyboardQmlViewStep( QObject* parent )
+    : Calamares::QmlViewStep(parent )
+    , m_config( new Config(this) )
+    , m_nextEnabled( false )
+    , m_writeEtcDefaultKeyboard( true )
+{
+    m_config->init();
+    m_nextEnabled = true;
+    emit nextStatusChanged( m_nextEnabled );
+}
+
+void
+KeyboardQmlViewStep::classBegin()
+{
+}
+
+void
+KeyboardQmlViewStep::componentComplete()
+{
+//     m_config->onActivate();
+}
+
+QString
+KeyboardQmlViewStep::prettyName() const
+{
+    return tr( "Keyboard" );
+}
+
+
+QString
+KeyboardQmlViewStep::prettyStatus() const
+{
+    return m_prettyStatus;
+}
+
+bool
+KeyboardQmlViewStep::isNextEnabled() const
+{
+    return m_nextEnabled;
+}
+
+bool
+KeyboardQmlViewStep::isBackEnabled() const
+{
+    return true;
+}
+
+bool
+KeyboardQmlViewStep::isAtBeginning() const
+{
+    return true;
+}
+
+bool
+KeyboardQmlViewStep::isAtEnd() const
+{
+    return true;
+}
+
+QList< Calamares::job_ptr >
+KeyboardQmlViewStep::jobs() const
+{
+    return m_jobs;
+}
+
+void
+KeyboardQmlViewStep::onActivate()
+{
+    m_config->onActivate();
+}
+
+void
+KeyboardQmlViewStep::onLeave()
+{
+    m_config->finalize();
+    m_jobs = m_config->createJobs( m_xOrgConfFileName,
+                                   m_convertedKeymapPath,
+                                   m_writeEtcDefaultKeyboard );
+    m_prettyStatus = m_config->prettyStatus();
+}
+
+QObject *
+KeyboardQmlViewStep::getConfig()
+{
+    return m_config;
+}
+
+void
+KeyboardQmlViewStep::setConfigurationMap( const QVariantMap& configurationMap )
+{
+    if ( configurationMap.contains( "xOrgConfFileName" ) &&
+            configurationMap.value( "xOrgConfFileName" ).type() == QVariant::String &&
+            !configurationMap.value( "xOrgConfFileName" ).toString().isEmpty() )
+    {
+        m_xOrgConfFileName = configurationMap.value( "xOrgConfFileName" )
+                             .toString();
+    }
+    else
+        m_xOrgConfFileName = "00-keyboard.conf";
+
+    if ( configurationMap.contains( "convertedKeymapPath" ) &&
+            configurationMap.value( "convertedKeymapPath" ).type() == QVariant::String &&
+            !configurationMap.value( "convertedKeymapPath" ).toString().isEmpty() )
+    {
+        m_convertedKeymapPath = configurationMap.value( "convertedKeymapPath" )
+                                .toString();
+    }
+    else
+        m_convertedKeymapPath = QString();
+
+    if ( configurationMap.contains( "writeEtcDefaultKeyboard" ) &&
+            configurationMap.value( "writeEtcDefaultKeyboard" ).type() == QVariant::Bool )
+        m_writeEtcDefaultKeyboard = configurationMap.value( "writeEtcDefaultKeyboard" ).toBool();
+    else
+        m_writeEtcDefaultKeyboard = true;
+
+    Calamares::QmlViewStep::setConfigurationMap( configurationMap ); // call parent implementation last
+    setContextProperty( "Keyboard", m_config );
```

Comment:
Why are individual modules setting a context property for their own config? The QmlViewStep already sets the (rather generic) name *Config* to the config object from each module.

Reply:
i thought it was needed to be called ? i saw that done in the welcomeqmlviewstep